### PR TITLE
[storage][cuda] Add cuda backend storage implementation

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

We will achieve a fully support for running on GPUs when reaches Mars 1.0, cuda storage backend is the first step for this goal.

### CODE REVIEW

File: mars/storage/cuda.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+except ImportError:
+    MemoryPointer = None
+    Buffer = None
+
+cupy = lazy_import('cupy', globals=globals())
+cudf = lazy_import('cudf', globals=globals())
+
+
+class CudaObjectId:
+    def __init__(self, headers: Dict, ptrs: List[int]):
```
Comment (qinxuye) [inline mars/storage/cuda.py:40]: Add `__slots__`, and make headers and ptrs public.

File: mars/storage/cuda.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        ptrs = object_id.ptrs
+        data_type = headers.pop('data_type')
+        if data_type == 'cupy':
+            ptr = ptrs[0]
+            size = headers['lengths'][0]
+            device = headers.get('device', None)
+            cuda_buf = DeviceBuffer(ptr=ptr, size=size)
+            cupy_mem_pointer = MemoryPointer(
+                UnownedMemory(ptr, size, cuda_buf, device_id=device), 0)
+            return cupy.ndarray(
```
Comment (qinxuye) [inline mars/storage/cuda.py:158]: Should use `serializations.deserialize` as well.

**Review by qinxuye — COMMENTED** [review]
LGTM overall.

**Comment by qinxuye** [comment]
Besides, coverage of `storage/cuda.py` could be provided.

**Comment by hekaisheng** [comment]
I have ran the cases and the coverage of `storage/cuda.py` is up to the target number.

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**